### PR TITLE
Hosting startup updates for 3.0

### LINKED ISSUE #12864: Update Hosting Startup (plat-specific config) for 3.0
Web Host --&gt; Generic Host for 3.0

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 613ad64e-bc62-edf7-2e86-42210de718f4
* Version Independent ID: aabbbc32-3237-5ef9-c422-892dc6884a22
* Content: [Use hosting startup assemblies in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-2.2)
* Content Source: [aspnetcore/fundamentals/host/platform-specific-configuration.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/fundamentals/host/platform-specific-configuration.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #12864

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-2.1&branch=pr-en-us-14607)

* I don't think it's too crazy far off ........ because it ✨ *Just Works!*:tm: ✨. However, there are bound to be some *RexHacks*:tm: in there that we'll need to factor out.

* Because the *build.ps1* script in the RuntimeStore project creates an *additionalDeps* folder/file structure that doesn't match the folder/file layout on an actual machine, I add a path example in the text. For good measure, I added an example for the store location, too.

### CODE REVIEW

**Comment by guardrex** [comment]
ping @pakrym ... or can u suggest someone available?

**Comment by pakrym** [comment]
cc @anurse as I changed teams

**Comment by Rick-Anderson** [comment]
@jkotalik please review.

**Comment by jkotalik** [comment]
What has changed from 2.2 to 3.0 @guardrex, this is way too large for me to review 😄

**Comment by guardrex** [comment]
Yeah ... I think one of the things we're noticing with these "whole topic versioning" treatments is that engineering, **_you_**, can't see the updates clearly ... thus can't review updates in a timely fashion. That **won't change** either. When we go to 3.1 and then duplicate the entire 3.0 section and make small changes to it, we'll be back in the same situation.

I'll clarify exactly what changed in a comment here in the morning (Thursday morning) and then ping u back @jkotalik.

**Comment by Rick-Anderson** [comment]
> What has changed from 2.2 to 3.0 @guardrex, this is way too large for me to review 😄

You could skim through the [Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/host/platform-specific-configuration?view=aspnetcore-2.1&branch=pr-en-us-14607) - that's probably a good idea anyway.

**Comment by guardrex** [comment]
@jkotalik The changes are the following ...

* Converts relative API doc links to `xref` links.
* For the sample code that produces the additional deps, adds a `deployment` path segment to the location of the published additional deps to reflect that that's the folder in the project where they'll end up. Should make it a hair faster for the reader to locate in the sample app where these assets are ending up.
* The MSBuild `Reference` for `HostingStartupLibrary.dll` in the project file for a class lib assembly direct reference seems to want backslash-escaped backslashes in the path provided to `Include`, so I update that `<Reference>` `Include`:

  ```xml
  <ItemGroup>
    <Reference Include=".\\bin\\Debug\\netcoreapp2.1\\HostingStartupLibrary.dll">
      <HintPath>.\bin\Debug\netcoreapp2.1\HostingStartupLibrary.dll</HintPath>
      <SpecificVersion>False</SpecificVersion>
    </Reference>
  </ItemGroup>
  ```

* Notes that the sample code generates the `StartupDiagnostics` package in the *obj\packages* folder. Again, this is to help the reader find things after publishing their hosting startup.
* Add a line to indicate where the published runtime store assets will go on the target machine. The last sentence ("would later be moved" bit) is added ...

  > Generates the runtime store for `StartupDiagnostics` in the *store* folder. The `dotnet store` command in the script uses the `win7-x64` [runtime identifier (RID)](/dotnet/core/rid-catalog) for a hosting startup deployed to Windows. When providing the hosting startup for a different runtime, substitute the correct RID on line 37 of the script. The runtime store for `StartupDiagnostics` would later be moved to the user's or system's runtime store on the machine where the assembly will be consumed. The user runtime store install location for the `StartupDiagnostics` assembly is *.dotnet/store/x64/netcoreapp3.0/startupdiagnostics/1.0.0/lib/netcoreapp3.0/StartupDiagnostics.dll*.

  That only applies to the sample app and not when the env var points directly to the publish folder in the app ...

  ```powershell
  AppendToEnvironmentVariable "DOTNET_SHARED_STORE" "$PSScriptRoot\store\"
  ```

  The point is just to get something in here about **_where on a target system_** it goes.

* Same thing for the published `additionalDeps` for `StartupDiagnostics` ... provide a path on the target system where they go. The last sentence ("would later be moved" bit) is added ...

  > Generates the `additionalDeps` for `StartupDiagnostics` in the *additionalDeps* folder. The additional dependencies would later be moved to the user's or system's additional dependencies. The user `StartupDiagnostics` additional dependencies install location is *.dotnet/x64/additionalDeps/StartupDiagnostics/shared/Microsoft.NETCore.App/3.0.0/StartupDiagnostics.deps.json*.

  Same deal ... just want to get something about the target system in there. This only applies to the sample apps, and the env var for the sample apps are set up to consume this directly from the project's publish folder ...

  ```powershell
  AppendToEnvironmentVariable "DOTNET_ADDITIONAL_DEPS" "$PSScriptRoot\additionalDeps\"
  ```

  For this one and the prior path note for a target, we can work further on the language.

  Also note that those two paths shown above refer to `netcoreapp3.0` and `3.0.0`, but those do appear as `netcoreapp2.2` and `2.2.0` in the example paths of the 2.x version of the topic.

* New 3.0 sample apps confirmed working.

* Few other nit updates here and there.

I'll leave it to you, of course, if you want to review the full text.

On a larger point wrt whole-topic-versioning, it won't improve for engineering reviews going this way release-to-release. Notwithstanding the use of large versioning blocks (as it done here) or a version-by-file format approach, where the entire contents of the topic are placed into a new file, you'll be looking at *all new text every release* with a useless diff.

For something that had major engineering churn, it probably wouldn't matter either way. Recall how bad the Kestrel topic became when it flipped in and out of versioning blocks every few lines. 🙈

This whole-topic approach tho does kind'a lock you into it going forward. Certainly, a number of the topics are easier to update this way, but they become harder to review ... much harder sometimes ... and that's not going to change this way ... *forever* and *ever* .........

yoda
### *If once you start down the dark path, forever will it dominate your destiny. Consume you, it will!*

\- Yoda ([Frank Oz](https://www.frankoz.com/)) [Star Wars: Episode V: The Empire Strikes Back](https://www.starwars.com/), &copy;1980 [Lucasfilm Ltd.](https://www.lucasfilm.com/)/[The Walt Disney Company](https://www.thewaltdisneycompany.com/)

**Comment by Rick-Anderson** [comment]
@jkotalik  can you review this? I'm pushing to get file versioning. Once we have file versioning, you can use VS to diff the two files.

**Comment by Rick-Anderson** [comment]
cc @scottaddie  - we really need file versioning to make version updates reviewable.

File: aspnetcore/fundamentals/host/platform-specific-configuration.md [hunk 1]
```diff
@@ -5,22 +5,450 @@ description: Discover how to enhance an ASP.NET Core app from an external assemb
...
+* A hosting startup assembly is created from the console app without an entry point that:
+  * Includes a class that contains the `IHostingStartup` implementation.
+  * Includes a [HostingStartup](xref:Microsoft.AspNetCore.Hosting.HostingStartupAttribute) attribute to identify the `IHostingStartup` implementation class.
+* The console app is published to obtain the hosting startup's dependencies. A consequence of publishing the console app is that unused dependencies are trimmed from the dependencies file.
+* The dependencies file is modified to set the runtime location of the hosting startup assembly.
+* The hosting startup assembly and its dependencies file is placed into the runtime package store. To discover the hosting startup assembly and its dependencies file, they're listed in a pair of environment variables.
+
+The console app references the [Microsoft.AspNetCore.Hosting.Abstractions](https://www.nuget.org/packages/Microsoft.AspNetCore.Hosting.Abstractions/) package:
+
+[!code-xml[](platform-specific-configuration/samples-snapshot/3.x/StartupEnhancement.csproj)]
```
Comment (Tratcher) [inline aspnetcore/fundamentals/host/platform-specific-configuration.md]: The internal review build is showing all 2.1 contents for these snippets, but the actual files you're linking to look correct.

**Comment by jkotalik** [comment]
Should:

```
<ItemGroup>
  <Reference Include=".\\bin\\Debug\\netcoreapp2.1\\HostingStartupLibrary.dll">
    <HintPath>.\bin\Debug\netcoreapp2.1\HostingStartupLibrary.dll</HintPath>
    <SpecificVersion>False</SpecificVersion>
  </Reference>
</ItemGroup>
```

be netcoreapp3.0?

File: aspnetcore/fundamentals/host/platform-specific-configuration.md (hunk 1 above)
Comment (guardrex) [reply aspnetcore/fundamentals/host/platform-specific-configuration.md]: That's correct ... the build gives us hard links to the 2.1 version. As long as u can change the version and get the other versions, it should be ok. Idk why it does that. We can ask Scott if you want.

**Comment by guardrex** [comment]
@jkotalik I'll need to see exactly where. There should be two versions of that. I'll check.

File: aspnetcore/fundamentals/host/platform-specific-configuration.md (hunk 1 above)
Comment (Tratcher) [reply aspnetcore/fundamentals/host/platform-specific-configuration.md]: I did change to 3.1 and still saw 2.1 snippets.

Comment (Tratcher) [reply aspnetcore/fundamentals/host/platform-specific-configuration.md]: Ah, 3.0 works, it's just 3.1 that's falling back to 2.1.

**Comment by guardrex** [comment]
@jkotalik Yes ... it's correct. It's in both version blocks, each with the correct TFM.

File: aspnetcore/fundamentals/host/platform-specific-configuration.md (hunk 1 above)
Comment (guardrex) [reply aspnetcore/fundamentals/host/platform-specific-configuration.md]: Good 😅 ... yeah, since there's no 3.0 (*yet*), that makes sense.

**Pull request merged**